ls1intum/Artemis
# `Programming exercises`: Add custom build scripts for local continuous integration

### Checklist
#### General


#### Server

### Motivation and Context


This PR builds on [#7528](https://github.com/ls1intum/Artemis/pull/7528#pullrequestreview-1726733152) and adds basic support for custom build scripts in Local CI.

### Description

Instructors can define script actions in the client during exercise creation which are then persisted in the database. During the set up of the build job container, these actions are retrieved for the build script.

### Steps for Testing

**Only test on TS3, TS5 or legacy TS11. Or test locally with LocalVC Local CI run configuration**
Prerequisites:
- 1 Admin/Instructor

1. Open "Create Programming Exercise" form, fill all necessary fields and enable custom build plans
2. Click on the displayed action item
3. Add something in the editor that will not interfere with the testing tasks, e.g. `echo "Hello World"`
![image](https://github.com/ls1intum/Artemis/assets/38919977/777479fb-a772-45e4-bb52-0e8407d5a580)
4. Create exercise
5. Wait till template and solution builds complete and verify results. (Template 0 out of 13, Solution 13 out of 13)

_________________________

1. Do the same as above but now remove the action item
2. Add a new action item (don't change the code)
3. Create exercise
4. Verify that both builds fail

### Testserver States
> [!NOTE]
> These badges show the state of the test servers.
> Green = Currently available, Red = Currently locked

![](https://byob.yarr.is/ls1intum/Artemis/artemis-test1)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test2)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test3)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test4)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test5)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test6)
![](https://byob.yarr.is/ls1intum/Artemis/artemis-test9)

### Review Progress






#### Performance Review
#### Code Review
#### Manual Tests
